upload.py

```diff
@@ -33,55 +35,80 @@ def hrsize(size):
             return '%.0f%sB' % (size, suffix)
     return '%.0f%sB' % (size, 'Y')
 
+
+class HashingFile(file):
+    def __init__(self, file_path):
+        super(HashingFile, self).__init__(file_path, "w+b")
+        self.sha1 = hashlib.sha1()
+
+    def write(self, data):
+        self.sha1.update(data)
+        return file.write(self, data)
+
+    def get_hash(self):
+        return self.sha1.hexdigest()
+
+
+def getHashingFieldStorage(upload_dir):
+    class HashingFieldStorage(cgi.FieldStorage):
+        def make_file(self, binary=None):
+            self.open_file = HashingFile(os.path.join(upload_dir, self.filename))
+            return self.open_file
+
+        def get_hash(self):
+            return self.open_file.get_hash()
+
+    return HashingFieldStorage
+
+
 class Upload(webapp2.RequestHandler):
 
     def get(self):
         self.response.write('Simple uploader\n')
 
-    def post(self, filename='upload.dat'):
-        return self.put(filename)
+    def post(self):
+        return self.put()
+
+    def put(self):
+        before = resource.getrusage(resource.RUSAGE_SELF)
+        before_io = psutil.disk_io_counters()
 
-    def put(self, filename='upload.dat'):
         upload_source = '%s (%s)' % (self.request.user_agent, self.request.client_addr)
         log.debug('incoming upload from ' + upload_source)
-        if self.request.content_type == 'multipart/form-data':
-            filestream = None
-            # use cgi lib to parse multipart data without loading all into memory; use tempfile instead
-            # FIXME avoid using tempfile; processs incoming stream on the fly
+        log.debug('type: ' + self.request.content_type)
+
+        if (self.request.content_type == 'multipart/form-data'):
             fs_environ = self.request.environ.copy()
             fs_environ.setdefault('CONTENT_LENGTH', '0')
             fs_environ['QUERY_STRING'] = ''
```

Comment:
I'm only now getting fully into the code. On second thought, we probably do need these lines. Would be nice to move them to the ```cgi.FieldStorage``` subclass, though, so we don't need to deal with them in our code.